### ISSUE TITLE
broken with uglify's default config

### ISSUE THREAD

**Issue by thecotne** [videojs/http-streaming#176]
build was fixed by changing uglify config here https://github.com/videojs/http-streaming/commit/0793ccb5bec2628ac25089be55adb54f3272a7f9

but many projects have uglify with default config minifying whole bundle 

and this breaks videojs

i think we can find way to make this project compatible with uglify's typeofs optimization

making it compatible with uglify's default config

**Comment by gkatsev** [comment]
Thanks for opening this issue, made me look further into it.
I do agree that we should do what we can to support the default options of uglify, however, I think that uglify having typeofs on by default is shipping an unsafe option by default. It would be nice if they flipped the default. There's lots of code that uses `typeof` checks, particularly around `window` and `global` which is where we this this issue occur most commonly. In addition, a lot of tooling is built around adding such type checks for example, browserify adds it when using `global` in a module https://github.com/browserify/insert-module-globals/blob/master/index.js#L32-L37 (we use global and browserify in mux.js). Our web worker bundling plugin also has a similar construct https://github.com/gkatsev/rollup-plugin-bundle-worker/blob/master/src/workerhelper.js#L1

What we're going to look into doing:
- [ ] Our usage of the `global` package means that there's a typeofs construct that breaks, so, we should get rid of it or find an alternative with an typeofs-safe alternative
- [ ] Our web workers used `window` a lot instead of `self` and that work incidentally because of our usage of the `global` lib, we should remove usage of `window` and use `self` instead
- [ ] mux.js uses `global`, we can probably refer to `window` directly or just call `escape` without dereferencing it from an object

The first two probably need to happen in conjunction but the mux.js change can happen separately.

**Comment by thecotne** [comment]
i am not sure how `typeofs` option break's videojs 

but `typeofs` is not considered "unsafe" by uglify

unsafe options have `unsafe_` prefix like `unsafe_arrows`, `unsafe_comps` etc.

i was testing uglify's behaviour here https://skalman.github.io/UglifyJS-online/

and could not find way to break logical operations with typeofs

**Comment by gkatsev** [comment]
it's because it transforms constructs like `typeof foo !== 'undefined'` into `void 0 !== foo` which if `foo` isn't not a defined variable will break. This is particularly important because our checks have `window` inside instead of `foo` and that code runs in WebWorkers which do not have `window` defined, so, things break. However, by following the steps above should work around the issue. Going to be working on this tomorrow.

**Comment by gkatsev** [comment]
Going to look into opening an issue against uglify for this as well, at least to bring it up as an issue and open discussion.

**Comment by gkatsev** [comment]
Ah, finally got a reduced test case. I guess this mostly happens when using commonjs and a bundler

```js
(function(window) {
  var win = typeof window !== 'undefined' ? window : {};

  win.postMessage('foo');
})(this)
```
via
```sh
$ uglifyjs --compress --mangle -- test.js
```
produces
```js
!function(o){(void 0!==o?o:{}).postMessage("foo")}(this);
```

I'll open this as an issue against uglify for discussion while working on trying to reduce usage of constructs like that so it works.

**Comment by gkatsev** [comment]
Actually, that example isn't good enough 😭

**Comment by gkatsev** [comment]
- [ ] https://github.com/videojs/http-streaming/pull/182
- [ ] https://github.com/videojs/mux.js/pull/209

**Comment by gkatsev** [comment]
So, I think the two PRs above together fix this.

**Comment by thecotne** [comment]
i think this need to be changed as well https://github.com/videojs/video.js/blob/master/build/minify.js#L11-L21

```js
compress: {
  sequences: true,
  dead_code: true,
  conditionals: true,
  booleans: true,
  unused: true,
  if_return: true,
  join_vars: true,
  drop_console: true,
  typeofs: false
}
```

everything except  `drop_console` and `typeofs` is default value and need to be removed to avoid confusion 

and `typeofs` need to be remove to avoid having bugs when minimized with uglify's default config

**Comment by gkatsev** [comment]
That shouldn't affect bundlers though because that minification only runs on the for the UMD builds of Video.js (though, our UMD builds are more like global builds but that's another matter). Video.js itself exposes a CJS and ES modules for bundler usage and that code is unminified.

**Comment by gkatsev** [comment]
This should now be fixed in 1.2.2 https://github.com/videojs/http-streaming/commit/2c42e8f6db175814ee11e3e3f6cfc596d1e39a9a, it'll get in to Video.js soon, hopefully.
